src/lib.rs

```diff
@@ -39,30 +42,53 @@ async fn start(
     // on the next call to get the next portion of objects (pagination)
     let mut continuation_token: Option<String> = None;
 
-    // Continuosly get the list of block data from S3 and stream send them to the `file_sink`
+    // Continuously get the list of block data from S3 and send them to the `file_sink`
     loop {
         // TODO: decide what to do if we got an error
         if let Ok(list_object_response) = s3_fetchers::list_blocks(
             &client,
             &bucket,
-            start_from_block_height.map(|block_height| block_height.to_string()),
+            Some(start_from_block_height.to_string()),
             continuation_token.clone(),
         )
         .await
         {
             // update the token for the next iter (pagination)
             continuation_token = list_object_response.continuation_token;
 
-            // read each of the block separately from S3
-            for folder in list_object_response.folder_names {
-                let streamer_message_json =
-                    s3_fetchers::get_object(&client, &bucket, &folder, &tracked_shards)
-                        .await
-                        .unwrap(); // TODO: handle error avoid unwraps
-                let streamer_message: near_indexer_primitives::StreamerMessage =
-                    serde_json::from_value(streamer_message_json).unwrap(); // TODO: handle error avoid unwraps
-                file_sink.send(streamer_message).await.unwrap(); // TODO: handle error avoid unwraps
-            }
+            let mut blocks_futures: futures::stream::FuturesOrdered<_> = list_object_response
+                .folder_names
+                .iter()
+                .map(|folder| {
+                    build_and_send_streamer_message(&file_sink, &client, &bucket, &folder)
+                })
+                .collect();
+
+            while let Some(_response) = blocks_futures.next().await {}
+        } else {
+            tracing::error!(
+                target: LAKE_FRAMEWORK,
+                "Failed to list objects from bucket {}. Retrying...",
+                &bucket
+            );
         }
     }
 }
+
+/// Fetches the necessary data to build up the `StreamerMessage`
+/// and send it to the stream (`file_sink`)
+async fn build_and_send_streamer_message(
+    file_sink: &mpsc::Sender<near_indexer_primitives::StreamerMessage>,
+    client: &Client,
+    bucket: &str,
+    folder: &str,
+) {
+    let streamer_message_json = s3_fetchers::get_object(&client, &bucket, &folder)
+        .await
+        .unwrap();
+    if let Ok(streamer_message) =
+        serde_json::from_value::<near_indexer_primitives::StreamerMessage>(streamer_message_json)
```

Comment:
What is the correct way we might want?

Reply:
The simplest way that we can survive even beyond 1.0 release is to warn and retry forever.

More advanced, we should probably expose Prometeus metrics, so we can hook Prometeus/Grafana/etc to monitor and alert on issues instead of manually monitoring the servers.